File: Algorithms/src/medium/Jon and Valyrian Sword.swift
/*
// Sample code to perform I/O:

let name = readLine()                           // Reading input from STDIN
print("Hi, ", name!, ".\n", separator: "")      // Writing output to STDOUT

// Warning: Printing unwanted or ill-formatted data to output will cause the test cases to fail
*/

// Write your code here
import Foundation

// ---------- Fast Input ----------
final class FastScanner {
    private var data: [UInt8] = Array(FileHandle.standardInput.readDataToEndOfFile()) + [0]
    private var idx: Int = 0
    
    func readInt() -> Int {
        var num = 0
        var sign = 1
        while data[idx] == 10 || data[idx] == 13 || data[idx] == 32 { idx += 1 }
        if data[idx] == 45 { sign = -1; idx += 1 }
        while data[idx] >= 48 {
            num = num * 10 + Int(data[idx] - 48)
            idx += 1
        }
        return num * sign
    }
}

let scanner = FastScanner()

// ---------- Input ----------
let n = scanner.readInt()
let x = scanner.readInt()
let totalSpecial = 2 * x

var isSpecial = Array(repeating: 0, count: n + 1)
for _ in 0..<totalSpecial {
    let v = scanner.readInt()
    isSpecial[v] = 1
}

// adjacency list: (neighbor, edge index)
var adj = Array(repeating: [(Int, Int)](), count: n + 1)

for i in 1..<n {
    let u = scanner.readInt()
    let v = scanner.readInt()
    adj[u].append((v, i))
    adj[v].append((u, i))
}

// ---------- DFS ----------
var visited = Array(repeating: false, count: n + 1)
var bestValue: Int64 = -1
var bestIndex: Int = 0

func dfs(_ u: Int) -> Int64 {
    visited[u] = true
    var count: Int64 = Int64(isSpecial[u])
    
    for (v, idx) in adj[u] {
        if !visited[v] {
            let sub = dfs(v)
            let crossings = sub * Int64(totalSpecial - Int(sub))
            
            if crossings > bestValue || (crossings == bestValue && idx > bestIndex) {
                bestValue = crossings
                bestIndex = idx
            }
            count += sub
        }
    }
    return count
}

// Tree is connected, root anywhere (1 is safe)
_ = dfs(1)

// ---------- Output ----------
print(bestIndex)

File: Algorithms/src/medium/Jon and Valyrian Sword.rs
/*
// Sample code to perform I/O:

use std::io;
use std::io::prelude::*;

fn main() {
    let mut name = String::new();
    io::stdin().read_line(&mut name).unwrap();          // Reading input from STDIN
    println!("Hi, {}.", name);                          // Writing output to STDOUT
}

// Warning: Printing unwanted or ill-formatted data to output will cause the test cases to fail
*/

// Write your code here
use std::io::{self, Read};

fn main() {
    // Fast input
    let mut input = String::new();
    io::stdin().read_to_string(&mut input).unwrap();
    let mut it = input.split_whitespace();

    let n: usize = it.next().unwrap().parse().unwrap();
    let x: i64 = it.next().unwrap().parse().unwrap();
    let total_special = 2 * x;

    // Read special nodes
    let mut is_special = vec![0i64; n + 1];
    for _ in 0..total_special {
        let v: usize = it.next().unwrap().parse().unwrap();
        is_special[v] = 1;
    }

    // Read edges
    let mut adj = vec![Vec::<(usize, usize)>::new(); n + 1];
    for idx in 1..n {
        let u: usize = it.next().unwrap().parse().unwrap();
        let v: usize = it.next().unwrap().parse().unwrap();
        adj[u].push((v, idx));
        adj[v].push((u, idx));
    }

    let mut visited = vec![false; n + 1];
    let mut best_value: i64 = -1;
    let mut best_index: usize = 0;

    fn dfs(
        u: usize,
        adj: &Vec<Vec<(usize, usize)>>,
        visited: &mut Vec<bool>,
        is_special: &Vec<i64>,
        total_special: i64,
        best_value: &mut i64,
        best_index: &mut usize,
    ) -> i64 {
        visited[u] = true;
        let mut cnt = is_special[u];

        for &(v, idx) in &adj[u] {
            if !visited[v] {
                let sub = dfs(
                    v,
                    adj,
                    visited,
                    is_special,
                    total_special,
                    best_value,
                    best_index,
                );

                let crossings = sub * (total_special - sub);

                if crossings > *best_value
                    || (crossings == *best_value && idx > *best_index)
                {
                    *best_value = crossings;
                    *best_index = idx;
                }

                cnt += sub;
            }
        }
        cnt
    }

    dfs(
        1,
        &adj,
        &mut visited,
        &is_special,
        total_special,
        &mut best_value,
        &mut best_index,
    );

    println!("{}", best_index);
}
